contracts/CourtPresaleActivate.sol

```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
         require(_token == address(presale.contributionToken()), ERROR_WRONG_TOKEN);
 
         // move tokens to this contract
-        require(ERC20(_token).safeTransferFrom(_from, address(this), _amount), ERROR_TOKEN_TRANSFER_FAILED);
+        ERC20 token = ERC20(_token);
+        require(token.safeTransferFrom(_from, address(this), _amount), ERROR_TOKEN_TRANSFER_FAILED);
 
+        _buyAndActivate(_from, _amount, _token);
+
+        // refund leftovers if any
+        _refund(token);
+    }
+
+    /**
+    * @dev This function needs a previous approval on the external token used for the contributed amount.
+    *      It will pull the approved tokens, convert them in Uniswap to the presale contribution token,
+    *      convert the obtained tokens in the presale instance, and activate the converted tokens in the
+    *      jurors registry instance of the Aragon Court.
+    * @param _token Address of the external contribution token used
+    * @param _amount Amount of contribution tokens to be converted and activated
+    * @param _minTokens Minimum amount of presale contribution tokens obtained in Uniswap
+    * @param _minEth Minimum amount of ETH obtained in Uniswap (Uniswap internally converts first to ETH and then to target token)
+    * @param _deadline Transaction deadline for Uniswap
+    */
+    function contributeExternalToken(
+        address _token,
+        uint256 _amount,
+        uint256 _minTokens,
+        uint256 _minEth,
+        uint256 _deadline
+    )
+        external
+    {
+        require(_amount > 0, ERROR_ZERO_AMOUNT);
+
+        // move tokens to this contract
+        require(ERC20(_token).safeTransferFrom(msg.sender, address(this), _amount), ERROR_TOKEN_TRANSFER_FAILED);
+
+        // get the Uniswap exchange for the contribution token
+        address payable uniswapExchangeAddress = uniswapFactory.getExchange(_token);
+        require(uniswapExchangeAddress != address(0), ERROR_UNISWAP_UNAVAILABLE);
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapExchangeAddress);
+
+        // swap tokens
+        address contributionTokenAddress = address(presale.contributionToken());
+        ERC20 token = ERC20(_token);
+        require(token.safeApprove(address(uniswapExchange), _amount), ERROR_TOKEN_APPROVAL_FAILED);
+        uint256 contributionTokenAmount = uniswapExchange.tokenToTokenSwapInput(_amount, _minTokens, _minEth, _deadline, contributionTokenAddress);
+
+        // buy in presale
+        _buyAndActivate(msg.sender, contributionTokenAmount, contributionTokenAddress);
+
+        // refund leftovers if any
+        _refund(token);
+        _refund(contributionTokenAddress);
```

Comment:
Similar to above the above questions.

At this point, we should be guaranteed to use up all of the received `token`, but the contract may have received some of `token` by mistake / random. Perhaps we should adjust the refund with the contract's starting balance of `token` before it pulls it?

And for `contributionToken`, would we ever have a situation where the presale would only accept a portion of the contribution?

-------

I'm wondering if we might want to think of a different recovery mechanism.

Given that there's no way to really track randomly deposited amounts, perhaps we can have an owned escape hatch callable by the AN DAO or etc.

Reply:
> Similar to above the above questions.
> 
> At this point, we should be guaranteed to use up all of the received `token`, but the contract may have received some of `token` by mistake / random. Perhaps we should adjust the refund with the contract's starting balance of `token` before it pulls it?

The initial motivation was more along the lines of bugs/inaccuracies of the external contracts, see [this comment](https://github.com/aragonone/court-presale-activate/pull/3#discussion_r367345562), but yours is a good point too.

> And for `contributionToken`, would we ever have a situation where the presale would only accept a portion of the contribution?

Theoretically not.

> I'm wondering if we might want to think of a different recovery mechanism.
> 
> Given that there's no way to really track randomly deposited amounts, perhaps we can have an owned escape hatch callable by the AN DAO or etc.

You read my mind, haha, I proposed that in a response above.